### PR TITLE
chroot: move to clap from getopts

### LINKED ISSUE #1732: chroot: move from getopts to clap
Code is in:
https://github.com/uutils/coreutils/blob/master/src/uu/chroot/src/chroot.rs

Tests:
https://github.com/uutils/coreutils/blob/master/tests/by-util/test_chroot.rs (well, we should write some)

As example:
https://github.com/uutils/coreutils/commit/9081e120d6b866c94679d9e29c7bdb17ccccb4d0

### PR DESCRIPTION
**Pull request by ycd**
Hey @sylvestre, implemented the basic structure with clap, however still need tests etc. I'll ping when you when it is ready to review. 

Closes #1732

### CODE REVIEW

File: src/uu/chroot/Cargo.toml [hunk 1]
```diff
@@ -1,23 +1,23 @@
 [package]
-name = "uu_chroot"
-version = "0.0.4"
 authors = ["uutils developers"]
-license = "MIT"
 description = "chroot ~ (uutils) run COMMAND under a new root directory"
+license = "MIT"
```
Comment (sylvestre) [inline src/uu/chroot/Cargo.toml:4]: please don't share the order :)

**Comment by ycd** [comment]
The clap part works fine and compatible with GNU's chroot, but i'm quite not sure how to properly test this.

File: src/uu/chroot/src/chroot.rs [hunk 2]
```diff
@@ -10,50 +10,61 @@
...
              Userspec has higher preference than -u and/or -g",
-            "USER:GROUP",
+                )
+                .value_name("USER:GROUP"),
         )
-        .parse(args);
+        .get_matches_from(args);
 
-    if matches.free.is_empty() {
+    if matches.args.is_empty() {
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:67]: I think you can enforce that above: make sure that clap verifies that it isn't empty

File: src/uu/chroot/src/chroot.rs [hunk 3]
```diff
@@ -97,11 +119,11 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
...
-fn set_context(root: &Path, options: &getopts::Matches) {
-    let userspec_str = options.opt_str("userspec");
-    let user_str = options.opt_str("user").unwrap_or_default();
-    let group_str = options.opt_str("group").unwrap_or_default();
-    let groups_str = options.opt_str("groups").unwrap_or_default();
+fn set_context(root: &Path, options: &clap::ArgMatches) {
+    let userspec_str = options.value_of("userspec");
+    let user_str = options.value_of("user").unwrap_or_default();
+    let group_str = options.value_of("group").unwrap_or_default();
+    let groups_str = options.value_of("groups").unwrap_or_default();
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:126]: would be nice to have a test to have proper coverage
(even if this is checking that it is failing)

File: src/uu/chroot/src/chroot.rs (hunk 2 above)
Comment (sylvestre) [reply src/uu/chroot/src/chroot.rs:67]: maybe like
https://github.com/uutils/coreutils/blob/master/src/uu/df/src/df.rs#L850-L857
?

**Review by sylvestre — COMMENTED** [review]
> i'm quite not sure how to properly test this.

at least test the error :)

**Review by sylvestre — CHANGES_REQUESTED** [review]
See the comment :)

File: src/uu/chroot/src/chroot.rs [hunk 4]
```diff
@@ -10,50 +10,61 @@
...
+                .value_name("USER"),
         )
-        .optopt("g", "group", "Group (ID or name) to switch to", "GROUP")
-        .optopt(
-            "G",
-            "groups",
-            "Comma-separated list of groups to switch to",
-            "GROUP1,GROUP2...",
+        .arg(
+            Arg::with_name("group")
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:42]: please declare options name like
https://github.com/uutils/coreutils/pull/1717/files#diff-8b30efe7f4221ac71ffa32926766963121d2494d4e3ff3e550b71960b0a94f5fR38

**Review by sylvestre — CHANGES_REQUESTED** [review]

File: src/uu/chroot/src/chroot.rs [hunk 5]
```diff
@@ -10,50 +10,69 @@
...
-            "USER",
+    let matches = App::new(executable!())
+        .version(VERSION)
+        .about(ABOUT)
+        .usage(SYNTAX)
+        .arg(Arg::with_name(options::NEWROOT).hidden(true))
+        .arg(
+            Arg::with_name(options::USER)
+                .short("u")
+                .long("user")
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:45]:
```suggestion
                .long(options::USER)
```

File: src/uu/chroot/src/chroot.rs [hunk 6]
```diff
@@ -10,50 +10,69 @@
...
+                .help("User (ID or name) to switch before running the program")
+                .value_name("USER"),
+        )
+        .arg(
+            Arg::with_name(options::GROUP)
+                .short("g")
+                .long("group")
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:52]:
```suggestion
                .long(options::GROUP)
```

File: src/uu/chroot/src/chroot.rs [hunk 7]
```diff
@@ -10,50 +10,69 @@
...
-        .optopt("g", "group", "Group (ID or name) to switch to", "GROUP")
-        .optopt(
-            "G",
-            "groups",
-            "Comma-separated list of groups to switch to",
-            "GROUP1,GROUP2...",
+        .arg(
+            Arg::with_name(options::GROUPS)
+                .short("G")
+                .long("groups")
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:59]:
```suggestion
                .long(options::GROUPS)
```

File: src/uu/chroot/src/chroot.rs [hunk 8]
```diff
@@ -10,50 +10,69 @@
...
+                .help("Comma-separated list of groups to switch to")
+                .value_name("GROUP1,GROUP2..."),
         )
-        .optopt(
-            "",
-            "userspec",
-            "Colon-separated user and group to switch to. \
+        .arg(
+            Arg::with_name(options::USERSPEC)
+                .long("userspec")
```
Comment (sylvestre) [inline src/uu/chroot/src/chroot.rs:65]:
```suggestion
                .long(options::USERSPEC)
```

File: src/uu/chroot/src/chroot.rs (hunk 2 above)
Comment (sylvestre) [reply src/uu/chroot/src/chroot.rs:75]: did you try this?
it should simplify the code

**Review by sylvestre — CHANGES_REQUESTED** [review]

**Comment by ycd** [comment]
> did you try this?
> it should simplify the code

Hmmm, sure, i can merge that with this line, since we have the exact logic below :+1: 

```rust
    let newroot: &Path = match matches.value_of(options::NEWROOT) {
        Some(v) => Path::new(v),
        None => crash!(1, "chroot: missing operand"),
    };
```

File: src/uu/chroot/src/chroot.rs (hunk 2 above)
Comment (ycd) [reply src/uu/chroot/src/chroot.rs:75]: I believe it is needed in his case since he uses `multiple(true)`.

Comment (sylvestre) [reply src/uu/chroot/src/chroot.rs:75]: You have min_values for this :)

example: https://github.com/uutils/coreutils/blob/a42e24e088a0ffa38af6549caa304c0cfcf4ac28/src/uu/realpath/src/realpath.rs#L61

File: tests/by-util/test_chroot.rs [hunk 9]
```diff
@@ -67,6 +67,11 @@ fn test_preference_of_userspec() {
     println!("result.stdout {}", result.stdout);
     println!("result.stderr = {}", result.stderr);
     let username = result.stdout.trim_end();
+
+    let result = scene.cmd("id -g -n").run();
```
Comment (ycd) [inline tests/by-util/test_chroot.rs:71]: CI fails on this
```
run: id -g -n
thread 'test_chroot::test_preference_of_userspec' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', tests/common/util.rs:562:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Comment (sylvestre) [reply tests/by-util/test_chroot.rs:71]: You can do it this way:
https://github.com/uutils/coreutils/blob/master/tests/by-util/test_whoami.rs#L41-L42

Comment (ycd) [reply tests/by-util/test_chroot.rs]: Awesome, thanks!

File: src/uu/chroot/src/chroot.rs (hunk 2 above)
Comment (sylvestre) [reply src/uu/chroot/src/chroot.rs]: it didn't work?

Comment (ycd) [reply src/uu/chroot/src/chroot.rs]: TBH, i wasn't sure which flag needs minimum value or multiple value, we use multiple values only for userspec and groups but they are seperated by colon or comma. Can you point me to the right direction if i'm missing smth :)

**Comment by ycd** [comment]
Any changes needed on this @sylvestre?

**Comment by sylvestre** [comment]
Excellent, thanks :)

**Pull request merged**